sql/sql_table.cc

```diff
@@ -9628,17 +9632,48 @@ fk_check_column_changes(THD *thd, Alter_info *alter_info,
       const bool equal_result= old_field->is_equal(*new_field);
       new_field->flags= flags;
 
-      if ((equal_result == IS_EQUAL_NO) ||
-          ((new_field->flags & NOT_NULL_FLAG) &&
-           !(old_field->flags & NOT_NULL_FLAG)))
+      if ((equal_result == IS_EQUAL_NO))
       {
         /*
           Column in a FK has changed significantly and it
           may break referential intergrity.
         */
+data_change:
         *bad_column_name= column->str;
         return FK_COLUMN_DATA_CHANGE;
       }
+      else if ((new_field->flags & NOT_NULL_FLAG) !=
+               (old_field->flags & NOT_NULL_FLAG))
+      {
+        if (referenced)
+        {
+          if (!(new_field->flags & NOT_NULL_FLAG))
+          {
+            /* Don't allow referenced column to change from
+            NOT NULL to NULL when foreign key relation is
+            UPDATE CASCADE and foreign key column doesn't allow NULL */
```

Comment:
The phrase "foreign key column" can refer to "referenced column" or  "referencing column". Let’s make it easier for the reader and end this with "ON UPDATE CASCADE and the referencing column is declared NOT NULL" .